File: go/leetcode/hard/41-缺失的第一个正数.go
package main

import (
	"fmt"
	"sort"
)

func firstMissingPositive(nums []int) int {
	sort.Ints(nums)
	temp := 1
	tempNum := 0
	for _, num := range nums {
		if num <= 0 {
			continue
		} else {
			if temp != num {
				if tempNum > 0 && tempNum != num {
					return temp
				}
			} else {
				temp++
			}
			tempNum = num
		}
	}
	return temp
}

func main() {
	nums := []int{0, 2, 2, 1, 1}
	fmt.Println(firstMissingPositive(nums))
}

File: python/leetcode/algorithms/hard/41-缺失的第一个正数.py
from typing import List


class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        nums = sorted(set(nums))
        temp = 1
        for i, num in enumerate(nums):
            if num <= 0:
                continue
            else:
                if temp != num:
                    return temp
                if i == len(nums) - 1:
                    return num + 1
                temp += 1
        return temp


nums = [0, 2, 2, 1, 1]
s = Solution()
print(s.firstMissingPositive(nums))
